autokeras/hypermodel/preprocessor.py

```diff
@@ -144,65 +144,156 @@ def set_weights(self, weights):
+                continue
+        for element in set4sentence:
+            if element not in self.sentence_containers:
+                self.sentence_containers[element] = 1
+            else:
+                self.sentence_containers[element] += 1
+        set4sentence.clear()
 
     def finalize(self):
-        self._texts = np.array(self._texts)
-        self.vectorizer.fit(self._texts)
-        data = self.vectorizer.transform(self._texts)
-        self.shape = data.shape[1:]
-        if self.targets:
-            self.selector = feature_selection.SelectKBest(
-                feature_selection.f_classif,
-                k=min(self.vectorizer.max_features, data.shape[1]))
-            self.selector.fit(data, self.targets)
+        for word in self.feature_counter:
+            _tf = self.feature_counter[word] / self.word_sum
+            idf = np.log(self.stc_num / self.sentence_containers[word] + 1)
+            self.tf_idf_vec[word] = _tf * idf
+        if len(self.vocabulary) < self._max_features:
+            self._max_features = len(self.vocabulary)
+        kbestfeature = dict(sorted(self.tf_idf_vec.items(),
+                                   key=lambda item: item[1],
+                                   reverse=True)[0:self._max_features])
+        self.kbestfeature_value = np.array(list(dict(
+            sorted(kbestfeature.items())).values()))
+        self.kbestfeature_key = np.array(list(dict(
+            sorted(kbestfeature.items())).keys()))
+        self.mask = np.zeros(self._max_features, dtype=int)
+        self._shape = np.shape(self.mask)
 
     def transform(self, x, fit=False):
-        sentence = nest.flatten(x)[0].numpy().decode('utf-8')
-        data = self.vectorizer.transform([sentence]).toarray()
-        if self.selector:
-            data = self.selector.transform(data).astype('float32')
-        return data[0]
+        x = nest.flatten(x)[0].numpy().decode('utf-8')
+        token_pattern = re.compile(r"(?u)\b\w\w+\b")
+        tokens = self._word_ngram(token_pattern.findall(x.lower()), self.ngram_range,
+                                  self.stop_words)
+        for feature in tokens:
+            try:
+                feature_idx = self.vocabulary[feature]
+                self.temp_vec.add(feature_idx)
+            except KeyError:
+                # Ignore out-of-vocabulary items for fixed_vocab=True
+                continue
+        for num in self.temp_vec:
+            try:
+                num = list(self.kbestfeature_key).index(num)
+            except ValueError:
+                continue
+            if num - 1 <= self._max_features:
+                self.mask[num] = 1
+
+        self.result = self.mask * self.kbestfeature_value
+        # Refresh the mask&temp_vec for next time usage.
+        self.mask = np.zeros(self._max_features, dtype=int)
+        self.temp_vec = set()
+        # TODO: For each x, what is the type of return value?
```

Comment:
The output should be an eager tensor. The input x is also an eager tensor. The self.result is a local variable. Is it used anywhere else? if not we can make it local.

Reply:
So should I use the tf.convert_to_tensor on the numpy array being returned?

Reply:
No, it should be converted automatically. The current implementation is good.